### PR TITLE
[CosmosDB] Introducing the apiVersion 04-01-2020 

### PR DESCRIPTION
**Pull request by MehaKaushik**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by MehaKaushik** [comment]
@markcowl the CI is failing because of test failure resulting from breaking change. Can we override this?  I don't find the point of creating a PR in sdk for net repo, to get this checked in. Otherwise please suggest a better alternative.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+          "200": {
+            "description": "Successfully retrieved notebook workspace resources.",
+            "schema": {
+              "$ref": "#/definitions/NotebookWorkspaceListResult"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../stable/2019-08-01/cosmos-db.json#/definitions/ErrorResponse"
```
Comment (wmengmsft) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json:71]: can we point to 2020-03-01 for other type references in this file?

Comment (wmengmsft) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json:71]: same comment for 2020-04-01's notebook.json

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-05-04/notebook.json [hunk 2]
```diff
@@ -0,0 +1,461 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Cosmos DB",
+    "description": "Azure Cosmos DB Database Service Resource Provider REST API",
+    "version": "2020-05-04"
```
Comment (shurd) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-05-04/notebook.json:6]: I believe this is version "2020-04-01" in our RP? This is how the generated SDKs get API version, so they will see "api version does not exist"

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-05-04/examples/CosmosDBDatabaseAccountCreateMax.json [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    "resourceGroupName": "rg1",
+    "api-version": "2020-05-04",
+    "subscriptionId": "subid",
+    "createUpdateParameters": {
+      "location": "westus",
+      "tags": {},
+      "kind": "GlobalDocumentDB",
+      "properties": {
+        "databaseAccountOfferType": "Standard",
+        "ipRules": [
```
Comment (wmengmsft) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-05-04/examples/CosmosDBDatabaseAccountCreateMax.json:13]: should be "ipRules": [
 { "ipAddressOrRange": "10.10.10.10" },
{ "ipAddressOrRange": "12.24.48.0/10" }
]

**Review by wmengmsft — COMMENTED** [review]
I think we are missing cosmosdbcassandratablecreateupdate in the 2020-04-01 examples

**Comment by MehaKaushik** [comment]
@markcowl please review

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json [hunk 4]
```diff
@@ -0,0 +1,7517 @@
...
+          "$ref": "#/definitions/ShardKeys"
+        },
+        "indexes": {
+          "description": "List of index keys",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/MongoIndex"
+          }
+        },
+        "analyticalStorageTimeToLive": {
```
Comment (shurd) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:6503]: I believe these should all be "analyticalStorageTtl". In Constants.cs, the string value of the property differs from the name.

**Review by chiragg4u — COMMENTED** [review]
What is the reason for releasing Notebook related API's in older versions 2019-12-12 and 2020-03-01? Why this feature can not go in only the latest version 2020-04-01?

**Comment by wmengmsft** [comment]
> What is the reason for releasing Notebook related API's in older versions 2019-12-12 and 2020-03-01? Why this feature can not go in only the latest version 2020-04-01?
This is to fix earlier omissions. The notebook API was originally introduced in 2019-08-01, but was accidentally omitted in 2019-12-12 and 2020-03-01. This PR tries to bring it back into Swagger as the underlying APIs are supported for these versions. Please let us know if this is okay.

**Comment by MehaKaushik** [comment]
@chiragg4u please merge this at your earliest convenience today. Thanks

**Comment by chiragg4u** [comment]
>             "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"

Thanks for referring to common schema. :)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:48 in ed2d6ea. [](commit_id = ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c, deletion_comment = False)

**Comment by chiragg4u** [comment]
>         "ipRules": {

Looks like you are renaming the property from IPRangeFilter. Will this cause customers using this name in their resource queries for e.g. from Azure Resource Graph?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:5097 in ed2d6ea. [](commit_id = ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c, deletion_comment = False)

**Comment by wmengmsft** [comment]
> > ```
> >     "ipRules": {
> > ```
> 
> Looks like you are renaming the property from IPRangeFilter. Will this cause customers using this name in their resource queries for e.g. from Azure Resource Graph?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:5097 in [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c). [](commit_id = [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c), deletion_comment = False)

This is an intended breaking change in this API version. The old property is a single string containing all IP address ranges that are comma separated. It has a flaw that makes it hard for Azure policy evaluation. The new design, determined in conjunction with Azure Policy team, is making it an array of IpAddressOrRange objects. We don't want to keep the old property around in the new API version so customer does not get confused with 2 ways of doing the same thing.

I am not sure what is the impact on ARG. Any pointers will be appreciated and we can follow up.

**Comment by chiragg4u** [comment]
>             }

New field options are being added in many objects. Looks like its only used in GET calls. Please confirm.
Also lets mark it Readonly.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:4629 in ed2d6ea. [](commit_id = ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c, deletion_comment = False)

**Comment by chiragg4u** [comment]
>           "type": "string",

shouldn't this be integer as throughput?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:6008 in ed2d6ea. [](commit_id = ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c, deletion_comment = False)

**Comment by MehaKaushik** [comment]
> > ```
> >       "type": "string",
> > ```
> 
> shouldn't this be integer as throughput?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:6008 in [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c). [](commit_id = [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c), deletion_comment = False)

That is intended. We return Minimum throughput as string.

**Comment by MehaKaushik** [comment]
> > ```
> >         }
> > ```
> 
> New field options are being added in many objects. Looks like its only used in GET calls. Please confirm.
> Also lets mark it Readonly.
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json:4629 in [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c). [](commit_id = [ed2d6ea](https://github.com/Azure/azure-rest-api-specs/commit/ed2d6eac97eebf03bfeb5808501c3d8e8a361e5c), deletion_comment = False)

Yes, they are only in GET calls. Would make the object readonly.

**Comment by MehaKaushik** [comment]
@chiragg4u please have a look again and merge if there are no more comments

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+          }
+        }
+      },
+      "delete": {
+        "tags": [
+          "NotebookWorkspacesResource"
+        ],
+        "description": "Deletes the notebook workspace for a Cosmos DB account.",
+        "operationId": "NotebookWorkspaces_Delete",
+        "x-ms-long-running-operation": true,
```
Comment (chiragg4u) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json:185]: >"x-ms-long-running-operation": true, [](start = 8, length = 36)

delete is a long running operation but PUT not. Is it intentional?

**Comment by chiragg4u** [comment]
ARG [https://docs.microsoft.com/en-us/azure/governance/resource-graph/overview](https://docs.microsoft.com/en-us/azure/governance/resource-graph/overview) is being used in Azure Portal and many customers are using Azure Resource Graph Explorer to create/save queries ([https://docs.microsoft.com/en-us/azure/governance/resource-graph/tutorials/create-share-query](https://docs.microsoft.com/en-us/azure/governance/resource-graph/tutorials/create-share-query)) which would fail if some properties are being changed.

---
In reply to: [629480674](https://github.com/Azure/azure-rest-api-specs/pull/9302#issuecomment-629480674) [](ancestors = 629480674)

**Review by chiragg4u — APPROVED** [review]
Approved with minor comments.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json (hunk 5 above)
Comment (MehaKaushik) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/notebook.json:185]: if you see in line 132 here PUT, represented by CreateOrUpdate, is a long-running operation.

**Comment by MehaKaushik** [comment]
> ARG https://docs.microsoft.com/en-us/azure/governance/resource-graph/overview is being used in Azure Portal and many customers are using Azure Resource Graph Explorer to create/save queries (https://docs.microsoft.com/en-us/azure/governance/resource-graph/tutorials/create-share-query) which would fail if some properties are being changed.
> 
> In reply to: [629480674](https://github.com/Azure/azure-rest-api-specs/pull/9302#issuecomment-629480674) [](ancestors = 629480674)

For the older apiVersions we are still supporting the old property name, so none of the existing scenarios will fail.

**Comment by MehaKaushik** [comment]
@chiragg4u can you please merge the PR as well. It needs someone who has write access to the repo.

**Pull request merged**